### ISSUE TITLE
False positive `cyclic-import` with non top-level imports (3.x)

### ISSUE THREAD

**Issue by jonathanberthias** [pylint-dev/pylint#9124]
### Bug description

When I import a file from within a function, it gets incorrectly flagged as a `cyclic-import`.

To reproduce, create 2 files inside a folder `mod`:
File `mod/a.py`:
```python
import mod.b
```

File `mod/b.py`:
```python
def f():
    import mod.a
```

### Command used

```shell
pylint mod --disable=all --enable=cyclic-import
```

### Pylint output

```python
mod/b.py:1:0: R0401: Cyclic import (mod.a -> mod.b) (cyclic-import)
```

### Expected behavior

No error, as the `mod.a` import is not run when importing `mod.b`.

Pylint 2.x does not report this as an error.

### Pylint version

```shell
pylint 3.0.1
astroid 3.0.0
Python 3.11.5 (main, Aug 24 2023, 12:23:19) [GCC 11.4.0]
```

**Comment by Pierre-Sassoulas** [comment]
Agree with @sodul. A 'real' cyclic import is a runtime error that will be very easy to detect unless you don't have a single test importing this part if the code. It might even prevent pylint analysis entirely. So runtime import is actually the main / more useful use case.

**Comment by jonathanberthias** [comment]
I didn't realize this was a new/improved check, the release notes for 3.0.0 don't mention anything about it.
Thanks for confirming!
